ext/git.c

```diff
@@ -0,0 +1,308 @@
+#else
+    return last_forward_slash;
+#endif
+}
+
+zend_string *read_git_file(const char *path, bool persistent) {
+    FILE *file = fopen(path, "r");
+    if (!file) {
+        return NULL;
+    }
+
+    char buffer[256];
+    size_t len = fread(buffer, 1, sizeof(buffer) - 1, file);
+    fclose(file);
+
+    if (len == 0) {
+        return NULL;
+    }
+
+    buffer[len] = '\0';
+    len = remove_trailing_newline(buffer);
+    return zend_string_init(buffer, len, persistent);
+}
+
+zend_string *get_commit_sha(const char *git_dir) {
+    char head_path[PATH_MAX];
+    snprintf(head_path, sizeof(head_path), "%s/HEAD", git_dir);
+
+    zend_string *head_content = read_git_file(head_path, false);
+    if (!head_content) {
+        return NULL;
+    }
+
+    const char *ref_prefix = "ref: ";
+    if (strncmp(ZSTR_VAL(head_content), ref_prefix, strlen(ref_prefix)) == 0) {
+        char ref_path[PATH_MAX];
+        snprintf(ref_path, sizeof(ref_path), "%s/%s", git_dir, ZSTR_VAL(head_content) + strlen(ref_prefix));
+        zend_string_release(head_content);
+        return read_git_file(ref_path, true);
+    }
+
+    return head_content;
+}
+
+zend_string *get_repository_url(const char *git_dir) {
+    char config_path[PATH_MAX];
+    snprintf(config_path, sizeof(config_path), "%s/config", git_dir);
+
+    FILE *file = fopen(config_path, "r");
+    if (!file) {
+        return NULL;
+    }
+
+    char buffer[256];
+    zend_string *result = NULL;
+    const char *url_prefix = "url = ";
+
+    while (fgets(buffer, sizeof(buffer), file)) {
+        if (strncmp(buffer, "[remote \"origin\"]", 17) == 0) {
+            while (fgets(buffer, sizeof(buffer), file)) {
+                if (buffer[0] == '[') break;
+                char *url = strstr(buffer, url_prefix);
+                if (url) {
+                    result = zend_string_init(url + strlen(url_prefix), strlen(url) - strlen(url_prefix), 1);
+                    normalize_string(result);
+                    break;
+                }
+            }
+            break;
+        }
+    }
+
+    fclose(file);
+    return result;
+}
+
+zend_string *find_git_dir(const char *start_dir) {
+    char current_dir[PATH_MAX];
+    snprintf(current_dir, sizeof(current_dir), "%s", start_dir);
+
+    while (true) {
+        char git_dir[PATH_MAX];
+        snprintf(git_dir, sizeof(git_dir), "%s/.git", current_dir);
+
+        if (access(git_dir, F_OK) == 0) {
+            return zend_string_init(git_dir, strlen(git_dir), 0);
+        } else if (errno == EACCES || errno == EPERM) {
+            // If we don't have permission, assume we're in a git dir but can't access the metadata
+            return NULL;
+        }
+
+        char *last_slash = find_last_dir_separator(current_dir);
+        if (!last_slash) {
+            return NULL;
+        }
+
+        *last_slash = '\0';
+    }
+}
+
+void remove_credentials(zend_string *repo_url) {
```

Comment:
This must create a copy if `at != NULL`, to prevent possible race conditions and invalidating the global data (e.g. `get_DD_GIT_REPOSITORY_URL` may point to a persistent global)